crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+            choices.push(Self::arbitrary_custom_section);
+            if self.types.len() < self.config.max_types() {
+                choices.push(Self::arbitrary_type_section);
+            }
+
+            // TODO FITZGEN
+            //
+            // choices.push(Self::arbitrary_import_section);
+            // choices.push(Self::arbitrary_func_section);
+            // choices.push(Self::arbitrary_core_section);
+            // choices.push(Self::arbitrary_component_section);
+            // choices.push(Self::arbitrary_instance_section);
+            // choices.push(Self::arbitrary_export_section);
+            // choices.push(Self::arbitrary_start_section);
+            // choices.push(Self::arbitrary_alias_section);
+
+            let f = u.choose(&choices)?;
+            f(self, u)?;
+        }
+
+        // Ensure we've generated all of our minimums.
+        self.fill_minimums = true;
+        if self.types.len() < self.config.min_types() {
+            self.arbitrary_type_section(u)?;
+        }
+
+        Ok(())
+    }
+
+    fn arbitrary_custom_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        self.sections.push(Section::Custom(u.arbitrary()?));
+        Ok(())
+    }
+
+    fn arbitrary_type_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        self.sections
+            .push(Section::Type(TypeSection { types: vec![] }));
+
+        let min = if self.fill_minimums {
+            self.config.min_types().saturating_sub(self.types.len())
+        } else {
+            0
+        };
+
+        let max = self.config.max_types() - self.types.len();
+
+        arbitrary_loop(u, min, max, |u| {
+            let ty = self.arbitrary_type(u)?;
+            let section_idx = self.sections.len() - 1;
+            let section = match self.sections.last_mut().unwrap() {
+                Section::Type(section) => section,
+                _ => unreachable!(),
+            };
+            self.types.push((section_idx, section.types.len()));
+            section.types.push(ty.clone());
+            self.current_type_scope_mut().push(Rc::new(ty));
+            Ok(true)
+        })?;
+
+        Ok(())
+    }
+
+    fn arbitrary_type(&mut self, u: &mut Unstructured<'a>) -> Result<Type> {
+        match u.int_in_range::<u8>(0..=5)? {
+            0 => Ok(Type::Module(self.arbitrary_module_type(u)?)),
+            1 => Ok(Type::Component(self.arbitrary_component_type(u)?)),
+            2 => Ok(Type::Instance(self.arbitrary_instance_type(u)?)),
+            3 => Ok(Type::Func(self.arbitrary_func_type(u)?)),
+            4 => Ok(Type::Value(self.arbitrary_value_type(u)?)),
+            5 => Ok(Type::Compound(self.arbitrary_compound_type(u)?)),
+            _ => unreachable!(),
+        }
+    }
+
+    fn arbitrary_module_type(&mut self, u: &mut Unstructured<'a>) -> Result<ModuleType> {
+        // TODO: this currently only supports function type definitions,
+        // function imports, and function exports.
```

Comment:
It's a bit funky since module types have their own scope in the component model spec and this is different from the scopes in core Wasm (and the old core wasm module linking). And since the representations of these scopes are different, I think we would need some kind of trait to abstract the differences away while still dealing with things like type sizes, etc. It kinda seems like all that machinery would be more of a headache than some sorta-but-not-quite duplicated code here. Thoughts?

Reply:
Ah yeah I don't know for sure or anything like that that this will be nicely deduplicated with some core helper. I'm mostly wondering if there's a commonality between the imports of a core wasm module and the imports here which could get all these other kinds of imports "for free". Seems fine to just implement it and see what happens though.